backend/uclapi/roombookings/helpers.py

```diff
@@ -230,3 +248,60 @@ def how_many_seconds_until_midnight():
         day=tomorrow.day, hour=0, minute=0, second=0
     )
     return (midnight - datetime.datetime.now()).seconds
+
+
+def _get_bookings_for_room(bookings, siteid, roomid):
+    bookings_for_room = []
+    for booking in bookings:
+        if (
+            booking["roomid"] == roomid and
+            booking["siteid"] == siteid
+        ):
+            bookings_for_room.append(booking)
+    return bookings_for_room
+
+
+def overlaps(start1, end1, start2, end2):
+    """
+    takes 4 datetimes
+    checks if they overlap
+    """
+    return max(start1, start2) < min(end1, end2)
+
+
+def _filter_for_free_rooms(all_rooms, bookings, start, end):
+    """
+    Find all rooms which don't have any bookings.
+    Args:
+        all_rooms: All available rooms.
+        bookings: All the bookings made in the days of the given time period
+        start: Start time for free rooms
+        end: End time for free rooms
+    """
+    rooms_with_bookings = list(all_rooms)
+    for idx, room in enumerate(all_rooms):
+        bookings_for_room = _get_bookings_for_room(
+            bookings,
+            room["siteid"],
+            room["roomid"]
+        )
+        rooms_with_bookings[idx]["bookings"] = bookings_for_room
+
+    free_rooms = []
+    for room in rooms_with_bookings:
+        for booking in room["bookings"]:
+
+            booking_start = _localize_time(booking["start_time"])
```

Comment:
This could probably be done in `_get_bookings_for_room` to avoid looping over all the data again (unless I misunderstood something, which is entirely possible)